File: python/open_guardrail/guards/profanity_id.py
"""Detect Indonesian profanity."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_WORDS: list[str] = [
    "bangsat", "anjing", "babi", "kampret",
    "bajingan", "tolol", "bego", "goblok",
    "setan", "keparat",
]

_PATTERNS: list[re.Pattern[str]] = [
    re.compile(
        r"\b" + re.escape(w) + r"\b",
        re.IGNORECASE,
    )
    for w in _WORDS
]


class _ProfanityId:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "profanity-id"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        lower = text.lower()
        for word, pat in zip(_WORDS, _PATTERNS):
            if pat.search(lower):
                matched.append(word)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="profanity-id",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Indonesian profanity detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched": matched}
                if triggered
                else None
            ),
        )


def profanity_id(
    *, action: str = "block"
) -> _ProfanityId:
    return _ProfanityId(action=action)

File: packages/guards/src/profanity-id.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ProfanityIdOptions {
  action: 'block' | 'warn';
}

const WORDS: string[] = [
  'bangsat', 'anjing', 'babi', 'kampret', 'bajingan',
  'tolol', 'bego', 'goblok', 'setan', 'keparat',
];

export function profanityId(options: ProfanityIdOptions): Guard {
  return {
    name: 'profanity-id',
    version: '0.1.0',
    description: 'Indonesian profanity detection',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const matched: string[] = [];

      for (const word of WORDS) {
        const escaped = word.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const re = new RegExp(`\\b${escaped}\\b`, 'i');
        if (re.test(lower)) matched.push(word);
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'profanity-id',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(matched.length / 3, 1.0) : 0,
        message: triggered ? `Indonesian profanity detected: ${matched.slice(0, 3).join(', ')}` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Text contains Indonesian profanity' } : undefined,
      };
    },
  };
}
